### PR TITLE
Implement support for multiple text runs per shadow.

### PR DESCRIPTION
**Pull request by glennw**
This change is [Reviewable](https://reviewable.io/reviews/servo/webrender/1469)

### CODE REVIEW

**Comment by Gankro** [comment]
r+ on the api/wrench changes

File: webrender/src/frame_builder.rs [hunk 1]
```diff
@@ -106,6 +106,37 @@ pub struct FrameBuilderConfig {
...
+}
+
+impl PendingTextShadow {
+    fn new(shadow: TextShadow,
+           clip_and_scroll: ClipAndScrollInfo,
+           local_clip: &LocalClip) -> PendingTextShadow {
+        PendingTextShadow {
+            shadow: shadow,
+            text_primitives: Vec::new(),
+            clip_and_scroll: clip_and_scroll,
```
Comment (kvark) [inline webrender/src/frame_builder.rs]: nit: could be `clip_and_scroll,`

File: webrender/src/frame_builder.rs [hunk 2]
```diff
@@ -392,6 +428,37 @@ impl FrameBuilder {
...
+        let mut text_shadow = self.pending_text_shadows
+                                  .pop()
+                                  .expect("Too many PopTextShadows?");
+        if !text_shadow.text_primitives.is_empty() {
+            let prim_cpu = TextShadowPrimitiveCpu {
+                text_primitives: text_shadow.text_primitives,
+                shadow: text_shadow.shadow,
+            };
+
+            text_shadow.local_rect = text_shadow.local_rect
```
Comment (kvark) [inline webrender/src/frame_builder.rs]: no need to actually modify the `local_rect` here as opposed to just declaring a new variable (thus lifting `mut` on `text_shadow`)

File: webrender/src/prim_store.rs [hunk 3]
```diff
@@ -493,10 +499,36 @@ pub struct TextRunPrimitiveCpu {
...
+                          device_pixel_ratio: f32,
+                          display_list: &BuiltDisplayList) {
+        // Cache the glyph positions, if not in the cache already.
+        // TODO(gw): In the future, remove `glyph_instances`
+        //           completely, and just reference the glyphs
+        //           directly from the displaty list.
+        if self.glyph_instances.is_empty() {
+            let src_glyphs = display_list.get(self.glyph_range);
+            for src in src_glyphs {
+                self.glyph_instances.push(GlyphInstance {
```
Comment (kvark) [inline webrender/src/prim_store.rs]: nit: could look better with `self.glyph_instances.extend(..)`

File: webrender/src/prim_store.rs [hunk 4]
```diff
@@ -997,50 +1050,34 @@ impl PrimitiveStore {
...
-                                              &text.glyph_instances,
-                                              text.render_mode,
-                                              text.glyph_options);
+                // This is a text-shadow element. Create a render task that will
+                // render the text run to a target, and then apply a gaussian
+                // blur to that text run in order to build the actual primitive
+                // which will be blitted to the framebuffer.
+                let cache_width = (metadata.local_rect.size.width * device_pixel_ratio).ceil() as i32;
+                let cache_height = (metadata.local_rect.size.height * device_pixel_ratio).ceil() as i32;
+                let cache_size = DeviceIntSize::new(cache_width, cache_height);
```
Comment (kvark) [inline webrender/src/prim_store.rs]: could probably write via `(metadata.local_rect.size * device_pixel_ratio).ceil().to_i32()`

**Pull request merged**